### PR TITLE
Civetweb WebSocket Server sample

### LINKED ISSUE #25395: Websocket Server API
***WHAT***
Implementation of Websocket Server API would be helpful for embedded devices, which can be configured/controlled over Ethernet.

***WHY***
To such devices may belong Ethernet connected sound boxes, embedded routers or modems, toys, smart watches etc. Websocket Server API would allow configuration of such devices just right in your local network.

***CURRENT SITUATION***
Currently Zephyr has only [Websocket Client API](https://docs.zephyrproject.org/latest/reference/networking/websocket.html) implementation.

***SUGGESTED SOLUTION***
We may adapt [MinnowServer](https://github.com/RealTimeLogic/MinnowServer)  since it is under [GPL2 License](https://github.com/systemd/systemd/blob/master/LICENSE.GPL2) 

Or implement it same way as MinnowServer.

Any suggestions?

### LINKED ISSUE #27417: CivetWeb Enable WebSocket
**Is your enhancement proposal related to a problem? Please describe.**
Compiling **CivetWeb** for `nucleo_f767zi` board with `WebSocket` support is not possible.

Steps to reproduce are:
1. In `zephyrproject/modules/lib/civetweb/CMakeLists.txt` enable the option `CIVETWEB_ENABLE_WEBSOCKETS`
2. Compile the project as follows: `west build -p auto -b "nucleo_f767zi" "samples/net/sockets/civetweb"`
3. Following error should occur:
```
<some_your_path>/zephyrproject/modules/lib/civetweb/src/civetweb.c: In function 'websocket_client_thread':
<some_your_path>/zephyrproject/modules/lib/civetweb/src/civetweb.c:17311:19: error: storage size of 'sa' isn't known 17311 |  struct sigaction sa;
```

It seems to be, that there is no `signal.h` implementation for any `stm32`!

**Describe the solution you'd like**
Clear and proper project compilation with `CIVETWEB_ENABLE_WEBSOCKETS` option.

The file `<some_your_path>/zephyrproject/modules/hal/esp-idf/components/newlib/include/sys/signal.h` seems to have necessary structure. 

**Additional context**
According to above info I have following questions:
* Does anybody already up to integration of WebSockets within CivetWeb to Zephyr and/or stm32?
* Is it possible and/or necessary to adapt `<some_your_path>/zephyrproject/modules/hal/esp-idf/components/newlib/include/sys/signal.h` for stm32?

### LINKED ISSUE #28941: Civetweb: create separate directory
**Is your feature request related to a problem? Please describe.**
CivetWeb using samples have potential to spread over multiple directories (especially after merging of #28323).

**Describe the solution you'd like**
Put all CivetWeb using samples into one `civetweb` root directory. Thus allowing http/websocket and many more CivetWeb based samples existing in one root directory (e.g. `samples/net/civetweb/http_server`, `samples/net/civetweb/websocket_server`).

**Describe alternatives you've considered**
none. (Maybe your suggestion?)

**Additional context**
Should be done incrementally after merging of #28323 and  #30175

### PR DESCRIPTION
**Pull request by KozhinovAlexander**
Hello all. Here I am providing my PR with civetweb based WebSocket Server sample.

Please feel free to test.

Currently I have only one bug: jQuery can not be properly loaded from the HTTP server (***issue were solved***)

I will appreciate any helpful suggestions.

I've also adjusted buffers specifically for my example. Same way the issue #21179 can be solved. I will investigate the civetweb http sample later.

### CODE REVIEW

**Comment by Nukersson** [comment]
@jukkar How can I weit in TCP/IP or socket peotocoll in mutex manner in Zephyr?

**Comment by jukkar** [comment]
Hi @Nukersson, right now there are no maintainers for the civetweb module and relevant zephyr code. Would you be interested in maintaining civetweb?

**Comment by Nukersson** [comment]
> Hi @Nukersson, right now there are no maintainers for the civetweb module and relevant zephyr code. Would you be interested in maintaining civetweb?

Think you @jukkar for your offer. I am very proud to be asked for it. Can we Skype for about 10 Minutes? I have some more questions to you you to civetweb maintaining. Skyping would be fastest way to clarify them. You can drop me an eMail (see my signature in commits).

**Comment by jukkar** [comment]
@Nukersson about your big jQuery file issue, the PR #28302 should eventually fix it. There is now some problems in the version I just sent so I need to fix them before that PR is usable again.

**Comment by Nukersson** [comment]
@jukkar I've just looked through all the civetweb issues, listed in main repo. Some of them are from me :D This PR solves my issues and also points to the ways to solve huge memory consumption issues on current civetweb. After this PR is merged I will investigate memory consumption of civetweb http example. I think, I have good chances to reduce memory usage and show the ways to other developments hot to do it in their projects 👯

**Comment by Nukersson** [comment]
Can somebody take a look at README.rst build (see **Documentation GitHub Workflow / build (pull_request)**)? Somehow I can't get rid of the listed error since the error description is not helpful.

**Comment by MaureenHelm** [comment]
> Can somebody take a look at README.rst build (see **Documentation GitHub Workflow / build (pull_request)**)? Somehow I can't get rid of the listed error since the error description is not helpful.

@carlescufi can you take a look?

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (lochej) [inline samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:50]: I have nothing wrong personnaly with "identifier-less" function definition but checkpatch complains here ^^

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (lochej) [inline samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:37]: I'm also OK with this, I find it quite clean, but checkpatch reports this as a new typedef which is forbidden.

I would personnaly keep the typedef but maybe some others have objections to that ?

**Review by lochej — COMMENTED** [review]
Just a few notes regarding Checkpatch warnings.

File: samples/net/sockets/civetweb_websocket_server/README.rst
Comment (lochej) [inline samples/net/sockets/civetweb_websocket_server/README.rst:60]: Have you forgotten the back quotes here ? Could it cause your current check-warns error ?
```
 ``http://192.0.2.2:8080``
```

File: samples/net/sockets/civetweb_websocket_server/src/main.c
Comment (lochej) [inline samples/net/sockets/civetweb_websocket_server/src/main.c:20]: STRINIFY instead 👍

File: samples/net/sockets/civetweb_websocket_server/src/main.c
Comment (lochej) [inline samples/net/sockets/civetweb_websocket_server/src/main.c:37]: SRINGIFY() ?

```c
static const char * const options[] = {
		"listening_ports", STRINGIFY(HTTP_PORT),
		"num_threads", "1",
		"max_request_size", STRINGIFY(MAX_REQUEST_SIZE_BYTES),
		NULL
	};
```

File: samples/net/sockets/civetweb_websocket_server/README.rst
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/README.rst:60]: Thanks for pointing on it. I've already tested it. But I do it again.

File: samples/net/sockets/civetweb_websocket_server/src/main.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/main.c:20]: done

File: samples/net/sockets/civetweb_websocket_server/src/main.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/main.c:37]: done

**Comment by Nukersson** [comment]
> @Nukersson Thank you very much for this new sample ! It gives new opportunities for web functionnality with Zephyr 👍
> 
> I have been able to run it on my Nucleo_H743ZI board (no-stlink/custom) with direct ethernet connection to my PC.
> Just changed the config IP address to accomodate my ethernet static IP.
> 
> image
> 
> I'm only seeing something strange, when I hit **Refresh** on my Chrome web browser, it doesn't reload the page and loads indefinitely. The web page is still working though and I can still send/receive the echo from the board.
> image

Thank you very much for testing it. :)

I've found the reason for permanent loading after refresh. There is no handling in my javascript, which closes Websocket connection after refresh. Since I am not an expert in JS - I would leave it as is. Moreover, it does not stops this sample be a good possibility demonstrator from Zephyr's side.

**Comment by carlescufi** [comment]
> > Can somebody take a look at README.rst build (see **Documentation GitHub Workflow / build (pull_request)**)? Somehow I can't get rid of the listed error since the error description is not helpful.
> 
> @carlescufi can you take a look?

It seems like you introduced a Sphinx error:

```
##[error]/home/runner/work/zephyr/zephyr/doc/_build/rst/doc/samples/net/sockets/civetweb_websocket_server/README.rst:60: WARNING: 'any' reference target not found: Message Text
``

**Comment by Nukersson** [comment]
> > > Can somebody take a look at README.rst build (see **Documentation GitHub Workflow / build (pull_request)**)? Somehow I can't get rid of the listed error since the error description is not helpful.
> > 
> > @carlescufi can you take a look?
> 
> It seems like you introduced a Sphinx error:
> 
> ```
> ##[error]/home/runner/work/zephyr/zephyr/doc/_build/rst/doc/samples/net/sockets/civetweb_websocket_server/README.rst:60: WARNING: 'any' reference target not found: Message Text
> ``
> 

Looks like a Bingo. :) Any suggestions, how to solve it? Otherwise I will take a look at possible solution. Looks like master google has some ideas.

**Comment by carlescufi** [comment]
> > > > Can somebody take a look at README.rst build (see **Documentation GitHub Workflow / build (pull_request)**)? Somehow I can't get rid of the listed error since the error description is not helpful.
> > > 
> > > 
> > > @carlescufi can you take a look?
> > 
> > 
> > It seems like you introduced a Sphinx error:
> > ```
> > ##[error]/home/runner/work/zephyr/zephyr/doc/_build/rst/doc/samples/net/sockets/civetweb_websocket_server/README.rst:60: WARNING: 'any' reference target not found: Message Text
> > ``
> > ```
> 
> Looks like a Bingo. :) Any suggestions, how to solve it? Otherwise I will take a look at possible solution. Looks like master google has some ideas.

Not really, seems it's picking up the IP address for some reason, even though you enclosed it in ` `` `. Maybe look at other `.rst` files that also use IP addresses.

**Comment by Nukersson** [comment]
@carlescufi Thanks, this is a good advice. I will take a look.

File: samples/net/sockets/civetweb_websocket_server/README.rst
Comment (lochej) [inline samples/net/sockets/civetweb_websocket_server/README.rst:63]: Shouldn't it be 
```
``Message Text``
```

with 2 back quotes ? Looks like single back quotes are for references like zephyr files.

Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/README.rst]: You're right. That was the problem.

File: samples/net/sockets/civetweb_websocket_server/README.rst
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/README.rst]: Another advice from you (see it below) worked fine. Thanks.

File: samples/net/sockets/civetweb_websocket_server/src/external_mg_cry_internal_impl.inl
Comment (jukkar) [inline samples/net/sockets/civetweb_websocket_server/src/external_mg_cry_internal_impl.inl:9]: I would suggest using shorter module name as this makes log lines unnecessary long. Perhaps "mg_int" or similar would be more suitable here.

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (jukkar) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:37]: We have usually used typedefs for function pointers only in zephyr, so that is why the checkpatch complains.

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (jukkar) [inline samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:169]: I hope this gets declared as static, otherwise the stack consumption will be huge.

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (jukkar) [inline samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:186]: same here

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (jukkar) [inline samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:204]: and here

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c]: done

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:37]: @jukkar But using new `typedef` in a `*.c` false is safe, untill this `*.c` file is explicitly included somewhere, which is obviously bad programming behavior. Thus the warning for new `typedf` in `*.c` files is useless! Can you point on somebody, who is familiar with checkpath.pl to add this kind of warnings only for `*.h` files?

File: samples/net/sockets/civetweb_websocket_server/src/external_mg_cry_internal_impl.inl
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/external_mg_cry_internal_impl.inl:9]: I would prefer `mg_cry_internal_impl`and `LOG_ERR("%s @ %d in civetweb.c", STR_LOG_ALLOC(func), line);` in the function body. Thus the line length printed to the console is reduced and provides better and shorter description of error's root.

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:176]: The define of `__data_decl` is changed to `static` now. Thanks for advice!

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:193]: The define of `__data_decl` is changed to `static` now. Thanks for advice!

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:211]: The define of `__data_decl` is changed to `static` now. Thanks for advice!

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (jukkar) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:37]: > But using new `typedef` in a `*.c` false is safe

I would disagree. The reason that we try to avoid extra `typedef` is to help to read / review the code. Using typedef hides the actual type of the variable which sometimes is ok, but usually it is not useful or needed.

File: samples/net/sockets/civetweb_websocket_server/src/external_mg_cry_internal_impl.inl
Comment (jukkar) [reply samples/net/sockets/civetweb_websocket_server/src/external_mg_cry_internal_impl.inl:9]: Ok, no objections if you find the longer string useful.

File: samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:37]: > > But using new `typedef` in a `*.c` false is safe
> 
> I would disagree. The reason that we try to avoid extra `typedef` is to help to read / review the code. Using typedef hides the actual type of the variable which sometimes is ok, but usually it is not useful or needed.

I Luderstand it now: you following simplification paradigm. Although in my case defining new type is simplification itself. Therefore I see it as necessary in my implementation. If you agree I will leave my typedef at place.

Comment (jukkar) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:37]: > I understand it now. Although in my case defining new typedef is simplification itself. Therefore I see it as necessary in my implementation. If you agree I will leave my typedef at place.

I understand your point but still disagree. Having `struct http_handler foo` is more descriptive than `http_handler_t foo` for the casual reader of the code. I would really prefer than normal struct is used here instead of typedef.

Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/http_server_handlers.c:37]: Ok. I need then update `this_register_handlers` function since it uses an array of `http_handler_t`. Thus the structure itself will be eliminated. The update comes with next push.

File: samples/net/sockets/civetweb_websocket_server/src/libc_extensions.c
Comment (jukkar) [inline samples/net/sockets/civetweb_websocket_server/src/libc_extensions.c:1]: BTW, the libc_extensions.c file and the *.inl files are probably the same between various civetweb based applications. So should we re-use them between app instead of replicating the code.
I was thinking directory structure like (in the example below, the http_server would be the current civetweb sample application):
```
samples/net/sockets/civetweb/
	lib/
	http_server/
	websocket_server/
	....
```
What do you think?

Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/libc_extensions.c:1]: It is a good idea. I thought about it already. My opinion here is: First push all my current civetweb dependent PRs as they currently are, then putting it in separate civetweb folder with possible optimisations. I think, this will be possible until v2.5 release. Moreover, all the libc/minlibc functionalities may be changed with updated posix libc implementations.

Comment (jukkar) [reply samples/net/sockets/civetweb_websocket_server/src/libc_extensions.c:1]: Great, sounds like a plan!

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by Nukersson** [comment]
@erwango Can you test this PR on one of your boards?

**Comment by Nukersson** [comment]
@MaureenHelm @pfalcon @tbursztyka @pabigot Please provide review.

**Comment by xhpohanka** [comment]
Just few notes to code style (I do not want any flame war, just notes)

- Is there a real reason for `__code_decl` and `__data_decl`? Similar things in samples often confuses less skilled people and brings unnecessary complexity to codebases.
- Is there a reason for static functions forward declaration? By simple reordering you can avoid them and reduce code duplication.

File: samples/net/sockets/civetweb_websocket_server/README.rst
Comment (pfalcon) [inline samples/net/sockets/civetweb_websocket_server/README.rst:39]: Zephyr networking samples have standardized on 192.0.2.1 being the **device** address, and 192.0.2.2 being the (peer) **host** address. Here you have it reversed, which will lead to no end of confusion. Please fix it.

**Review by pfalcon — DISMISSED** [review]
Please fix reversed device/host address assignment.

File: samples/net/sockets/civetweb_websocket_server/README.rst
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/README.rst:39]: done

**Comment by Nukersson** [comment]
> Just few notes to code style (I do not want any flame war, just notes)
> 
> * Is there a real reason for `__code_decl` and `__data_decl`? Similar things in samples often confuses less skilled people and brings unnecessary complexity to codebases.

The idea behind this construct is providing a mechanism to put the code in different memory and easy change of its location at compile time. 
I think, that less skilled people can grow they skills with tutorials outside of Zephyr. Zephyr is not intended to provide C tutorial. Moreover, such examples may increase coding skills of less skilled people, cause not knowing such simple constructs, may trigger them to look for explanations and descriptions.

> * Is there a reason for static functions forward declaration? By simple reordering you can avoid them and reduce code duplication.

Yes there is. I am just avoiding reordering meant by you 😄 Sorry. It lets functions bodies be placed free inside of corresponding *c file, since the functions were previously forward declared. May be in terms of number of line reduction, reordering can help. But, really, I am just using long established code styling (it's all about [MISRA](https://rules.sonarsource.com/c/RSPEC-819)).

File: samples/net/sockets/civetweb_websocket_server/README.rst
Comment (pfalcon) [reply samples/net/sockets/civetweb_websocket_server/README.rst]: Thanks for update.

**Comment by jukkar** [comment]
Would there be anyone on review list willing to review this so this one could proceed, this PR has been cooking for a long time now? Thanks in advance for all the help!

File: samples/net/sockets/civetweb_websocket_server/src/main.c
Comment (lochej) [inline samples/net/sockets/civetweb_websocket_server/src/main.c:69]: Just some remark on english language here 😃 Noticed it when trying your PR on my Nucleo_H743ZI.

If only one server is started then please replace **were** by **was** and I'll be happy 👍 
```suggestion
	LOG_INF("WebSocket Server was started!");
```

**Review by lochej — CHANGES_REQUESTED** [review]

File: samples/net/sockets/civetweb_websocket_server/src/main.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/main.c]: Merci beaucoup. Terminé.

**Comment by Nukersson** [comment]
@MaureenHelm @PiotrZierhoffer  Can you please trigger responsible reviewers.

File: samples/net/sockets/civetweb_websocket_server/src/main.c
Comment (Nukersson) [reply samples/net/sockets/civetweb_websocket_server/src/main.c]: @lochej please provide review or approve this PR

**Review by lochej — APPROVED** [review]

**Pull request merged**
